### ISSUE TITLE
Assertion "Calling a function with a bad signature!" failed (from SPIRVPreTranslationLegalizer.cpp)

### ISSUE THREAD

**Issue by iliya-diyachkov** [KhronosGroup/LLVM-SPIRV-Backend#60]
After 906a4a8 (#52) 5 tests fail on `(i >= FTy->getNumParams() || FTy->getParamType(i) == Args[i]->getType()) && "Calling a function with a bad signature!"' assertion in llvm/lib/IR/Instructions.cpp:495:
```
llvm-intrinsics/umul.with.overflow.ll
transcoding/SpecConstantComposite.ll
AtomicCompareExchange.ll
instructions/nested-composites.ll
instructions/call-complex-function.ll
```
All originates from SPIRVPreTranslationLegalizer.cpp

**Comment by iliya-diyachkov** [comment]
In all these tests the issue is with type mismatch of the first arg of inserted spv intrinsic. The expected type is i32 but the type of the passed arg is composite:
```
llvm-intrinsics/umul.with.overflow.ll		spv_extractv	llvm_i32_ty != { i8, i1 } %umul
transcoding/SpecConstantComposite.ll		spv_store	llvm_i32_ty != %struct._ZTS3POD.POD %32
AtomicCompareExchange.ll			spv_extractv	llvm_i32_ty != { i32, i1 } %1
instructions/nested-composites.ll		spv_insertv	llvm_i32_ty != %ty2 undef (also composite type)
instructions/call-complex-function.ll		spv_insertv	llvm_i32_ty != { i32, i16 } undef
```
so we need to either allow any type of the first arg or insert additional cast instructions.

**Comment by zuban32** [comment]
Well, just noticed this. If the problem is still present (haven't checked) - in the initial design I proposed the idea for the composite usecases was that all the instructions that had composite operands should get i32s instead which are in turn created by the composite creation intrinisics. That's why IMO creating either casts or modifying the intrinsic argument types might be incorrect, the problem just seems like an incorrect pre-legalization. I'll try to provide more context once I take a look on these tests myself

**Comment by iliya-diyachkov** [comment]
@zuban32, the issue is still here. I attempted to solve it some time ago probably trying the way you have described too, but with no success, so I decided to switch to other issues for a while. I'm going to return to it at the end of the week and update its status. Perhaps the issue requires different solution.

For example there is IR from instructions/nested-composites.ll:
```
%ty2 = type { %ty1, i64 }
...
define i32 @producer(i16 %a, i32 %b, i64 %c) {
  %agg1 = insertvalue %ty2 undef, i16 %a, 0, 0
  %agg2 = insertvalue %ty2 %agg1, i32 %b, 0, 1
  %agg3 = insertvalue %ty2 %agg2, i64 %c, 1
  ret %ty2 %agg3
}
```
It fails on conversion of the first insertvalue to "declare i32 @llvm.spv.insertv.i16(i32 %0, i16 %1, ...)" since types of formal and actual parameters don't match (composite %ty2 vs i32).

As I remember the problem is we cannot pass an arg of composite type (even to our artificial intrinsic calls) without lowering it to a set of args of simple types. So I had an idea to introduce an intrinsic similar to our const_composite, but which builds a composite type from simple ones. 

Another way would be to hide somehow a composite type to metadata while IRTranslator works. But it did not work in my experiments.

**Comment by zuban32** [comment]
#76 makes a good progress on this, and especially for 'nested-composites'.

Answering your question - all of these insertvalues are to return i32s instead of %ty2, and since the corresponding intrinsic expects exactly that we're fine with it, the problem was to handle 'undef' in the first one properly.

There was also a problem regarding the funcs processing, but that's fixed in #76 as well.

It doesn't affect the passrate, but compilation of some tests makes much further with it.

**Comment by iliya-diyachkov** [comment]
Great, thanks!
Looking thru changed output for instructions/nested-composites.ll I see that type of that 'undef' and connected OpCompositeInserts is i32 (although the composite type is present in the file), so we need to correct it. Also indexation in the first and the second OpCompositeInserts wrong (should be "0 0" and "0 1" correspondingly):
```
        %2 = OpTypeInt 32 0
...
        %4 = OpTypeInt 64 0
        %5 = OpTypeInt 16 0
        %6 = OpTypeStruct %5 %2
        %7 = OpTypeStruct %6 %4
...
        %13 = OpUndef %2
...
        %25 = OpCompositeInsert %2 %20 %13 0
        %26 = OpCompositeInsert %2 %21 %25 0
        %27 = OpCompositeInsert %2 %22 %26 1
```

**Comment by zuban32** [comment]
Yes, and moreover the reason why the tests is still failing on compilation stage is wrong 'assign_type's generation - somehow we got a chains of them (i.e. ASSIGN_TYPE(ASSIGN_TYPE(Val, Ty1), Ty2))

**Comment by zuban32** [comment]
BTW @iliya-diyachkov you may assign this issue to me since I'm already looking into it

**Comment by iliya-diyachkov** [comment]
The issue is resolved in several commits (#181 #184 #191) for all tests except llvm-intrinsics/umul.with.overflow.ll. It will be fixed by adding new function in #82.
